Propagate callable parameter types using ParamSpec (#142306)

kaspell opened the pull request:
The codebase has a few locations where callable parameter type information is lost when the unpackings *args and **kwargs are typed as Any. Refactor these instances to retain type information using typing_extensions.ParamSpec.

Also, in these functions, enforce return type with TypeVar.

Addresses #142306

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @malfet @xuzhao9 @gramster @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

kaspell committed:
Propagate callable parameter types using ParamSpec (#142306)

The codebase has a few locations where callable parameter type
information is lost when the unpackings *args and **kwargs are
typed as Any. Refactor these instances to retain type information
using typing_extensions.ParamSpec.

Also, in these functions, enforce return type with TypeVar.

Skylion007 commented on torch/_dynamo/external_utils.py:
```diff
@@ -35,13 +37,13 @@ def is_compiling() -> bool:
 
 
-def wrap_inline(fn: Callable[..., Any]) -> Callable[..., Any]:
+def wrap_inline(fn: Callable[_P, Any]) -> Callable[_P, Any]:
```
Could you also propogate the return type too, yby replace 'Any' with a `_R = TypeVar("_R")`

kaspell replied:
Sure, will do!

Skylion007 commented on torch/utils/_sympy/functions.py:
```diff
@@ -32,6 +32,7 @@
 
 
 _T = TypeVar("_T", bound=SupportsFloat)
+_Ts = TypeVarTuple("_Ts")
```
Why is this suppose to be a TypeVarTuple? Are the inputs really suppose to be Types here?

kaspell replied:
This allows us to match an unspecified number of positional arguments of possibly different types and then pass that information to the nested function and onwards.

This was a workaround to a "too few arguments" mypy error [0] if one tried to use _P.args here. Another workaround was to (needlessly) add keyword unpacking to the wrapper signature and the function call, but I didn't think that was a good choice.

I can also undo this part if TypeVarTuple isn't standard practice in the codebase, if I've misunderstood something, or for some other reason! Let me know if so.

[0] Due to _P encompassing both positional and keyword arguments and the function call then missing keyword arguments.

From https://docs.python.org/3.11/library/typing.html#typing.TypeVarTuple:
<img width="814" alt="Screenshot 2024-12-25 at 22 38 55" src="https://github.com/user-attachments/assets/9e9c2df6-6661-47f3-be8a-149fb860a219" />

Skylion007 commented on torch/utils/_sympy/functions.py:
```diff
@@ -101,9 +102,9 @@ def _is_symbols_binary_summation(expr: sympy.Expr) -> bool:
 
 
-def _keep_float(f: Callable[..., _T]) -> Callable[..., Union[_T, sympy.Float]]:
+def _keep_float(f: Callable[[*_Ts], _T]) -> Callable[[*_Ts], Union[_T, sympy.Float]]:
```
'*' isn't supported on legacy Python versions, use typing_extensions.Unpack

kaspell replied:
Ah right, ty, will use Unpack! Also switched to using Optional in some other signatures for backwards compatibility.

Skylion007 commented on torch/utils/_sympy/functions.py:
```diff
@@ -101,9 +102,9 @@ def _is_symbols_binary_summation(expr: sympy.Expr) -> bool:
+def _keep_float(f: Callable[[*_Ts], _T]) -> Callable[[*_Ts], Union[_T, sympy.Float]]:
     @functools.wraps(f)
-    def inner(*args: Any) -> Union[_T, sympy.Float]:
+    def inner(*args: *_Ts) -> Union[_T, sympy.Float]:
```
```suggestion
    def inner(*args: Unpack[_Ts]) -> Union[_T, sympy.Float]:
```

kaspell replied:
Ty!

kaspell committed:
Use Unpack and Optional in signatures for backwards compatibility

The unpacking operator * and the union operator | don't exist in
some of the supported versions, so replace these with Unpack and
Optional, respectively.

Skylion007 commented on torch/distributed/_tools/fsdp2_mem_tracker.py:
```diff
@@ -185,8 +200,8 @@ def _instrument_fsdp_sharded_params_grads(
-        orig_fsdp_state_pre_fw: Callable,
-    ) -> Callable:
+        orig_fsdp_state_pre_fw: Callable[_P, Any],
+    ) -> Callable[_P, Tuple[Tuple[Any, ...], Dict[str, Any]]]:
```
Why tuple and dict any here?

kaspell replied:
That is just copied from the return type of the nested wrapper function which had that type.

Skylion007 replied:
Shouldn't this actually be the return type of `orig_fsdp_state_pre_fw`? That's what is actually returned (an the inner function should be accordingly updated...

kaspell replied:
It kind of is already – just with the quirk that the return value of `orig_fsdp_state_pre_fw` is unpacked in the inner function and then the tuple is returned. So in this case a plain use of e.g. `TypeVar` for the return value would raise an error.

But now with this comment I noticed that in this signature we could replace the current return type `Any` of `orig_fsdp_state_pre_fw` with this tuple type as well. We could also use `TypeVarTuple` as before to replace the `Any, ...` to preserve more information. How does that sound? The types would look as follows

    def _fsdp_state_pre_forward(
        self,
        fsdp_mod: FSDPModule,
        orig_fsdp_state_pre_fw: Callable[_P, Tuple[Tuple[Unpack[_Ts]], Dict[str, Any]]],
    ) -> Callable[_P, Tuple[Tuple[Unpack[_Ts]], Dict[str, Any]]]:
        @wraps(orig_fsdp_state_pre_fw)
        def inner(
            *args: _P.args, **kwargs: _P.kwargs
        ) -> Tuple[Tuple[Unpack[_Ts]], Dict[str, Any]]:
            ...
            args, kwargs = orig_fsdp_state_pre_fw(*args, **kwargs)
            ...
            return args, kwargs

        return inner

Skylion007 replied:
Assuming that works @kaspell, looks good to me.

Skylion007 reviewed (approved):
Looks good to me (once the changes discussed above are resolved).

kaspell committed:
Enhance argument callable typing in _fsdp_state_pre_forward

Instead of using a return value Any, give the return value its
actual form. Also make other types in _fsdp_state_pre_forward
consistent with this change.

Skylion007 commented on torch/utils/_stats.py:
```diff
@@ -3,17 +3,23 @@
     simple_call_counter[label] = prev + 1
 
-def count(fn: Callable[..., Any]) -> Callable[..., Any]:
+def count(fn: Callable[_P, _R]) -> Callable[..., _R]:
```
```suggestion
def count(fn: Callable[_P, _R]) -> Callable[_P, _R]:
```

Skylion007 committed:
Add one missing ParamSpec

Skylion007 commented on torch/utils/_stats.py:
```diff
@@ -3,17 +3,23 @@
     simple_call_counter[label] = prev + 1
 
-def count(fn: Callable[..., Any]) -> Callable[..., Any]:
+def count(fn: Callable[_P, _R]) -> Callable[_P, _R]:
```
This is unearthing a real error, but that doesn't need to be fixed in this PR. Feel free to and a type ignore at the error below.

kaspell replied:
Got it! Added that and a fix to a lint error about trailing spaces that came up after the ignore comment.

kaspell committed:
Ignore type error about mismatching signatures in inheritance chain

Also fix a lint error about trailing spaces that came up after
adding the type ignore.

Skylion007 commented:
@pytorchbot merge

Skylion007 commented on torch/_dynamo/testing.py:
```diff
@@ -407,9 +410,9 @@ def check_dynamic_shape_capture() -> bool:
 
 
-def _make_fn_with_patches(fn: Callable[..., _T], *patches: Any) -> Callable[..., _T]:
+def _make_fn_with_patches(fn: Callable[_P, _T], *patches: Any) -> Callable[_P, _T]:
```
A lot more decorators here that could have their typing improved btw

XuehaiPan commented on torch/_dynamo/external_utils.py:
```diff
@@ -61,15 +64,15 @@ def call_hook(
 
 
-def wrap_numpy(f: Callable[..., Any]) -> Callable[..., Any]:
+def wrap_numpy(f: Callable[_P, pytree.PyTree]) -> Callable[_P, pytree.PyTree]:
```
```suggestion
def wrap_numpy(f: Callable[_P, _R]) -> Callable[_P, _R]:
```

Typevar `_R` is the better one than `PyTree`.

Skylion007 replied:
@XuehaiPan Follow up PR? I think this caused typing issues somewhere with optree's typing system?

XuehaiPan replied:
`pytree.PyTree` is just an alias to `typing.Any`.

Skylion007 committed:
Try wrap_numpy proper return type

Co-authored-by: Xuehai Pan <XuehaiPan@outlook.com>

Skylion007 commented:
@pytorchbot merge

Skylion007 commented:
@pytorchbot merge
